### PR TITLE
Kill JointTyped

### PR DESCRIPTION
**Pull request by gchenfc**
This abandons future plans for Ball/Spherical/U-joints.  Reasons:
* Gazebo doesn’t really handle ball joints fully anyway, since the different backends handle them differently and some don’t support them at all, so Gazebo just doesn’t fully support them, e.g. you cannot actuate ball joints.
* Previously I thought these would be useful to simulate cable robot, but I realized there’s a much better way.
* These types of joints (e.g. wrist, hip) are implemented with multiple “revolute” joints anyway, and due to joint limits don’t typically run into strange “singularity issues” (e.g. all joint angles = pi) anyway.
* Lots of other people just implement them as multiple intersecting joints, e.g. Lee09mmt which Frank shared (I didn’t review this paper in detail)
* The ball joints etc make lots of peoples’ lives (e.g. yetong) a lot harder.
* Therefore, I proposed we kill the complicated joint stuff and just revert back to only screw-type joints (joint angle has type double)

Frank’s response (after much deliberation):
* Ok
* Kill #286
* Gerry handle this new PR:
  * Kill big joint inheritance heirarchy
    * Basically, ScrewJointBase is the new Joint
  * Remove all joint templating madness
  * R-P-H joints only (H is ScrewJoint)
  * Minimal possible change to factors to make review easy

Can re-target to master if #290 is rejected.  I forget why I based this branch on the other one - probably just forgot to checkout back to master before starting this work.

### CODE REVIEW

File: python/tests/test_factors.py [hunk 1]
```diff
@@ -30,9 +30,7 @@ def setUp(self):
...
         """Test RevoluteJointMeasurementFactor."""
-        factor = gtd.RevoluteJointMeasurementFactor(
-            self.wTp_key, self.wTc_key,
-            gtsam.noiseModel.Isotropic.Sigma(6, 0.1),
-            self.robot.joint("joint_1"), np.pi / 4, self.k)
+        factor = gtd.JointMeasurementFactor(gtsam.noiseModel.Isotropic.Sigma(6, 0.1),
+                                            self.robot.joint("joint_1"), np.pi / 4, self.k)
 
-        self.assertTrue(isinstance(factor, gtd.RevoluteJointMeasurementFactor))
+        self.assertTrue(isinstance(factor, gtd.JointMeasurementFactor))
```
Comment (dellaert) [inline python/tests/test_factors.py:36]: So one factor rules them all? Awesome!

**Review by dellaert — CHANGES_REQUESTED** [review]
One more iteration. Let’s discuss #290 on Monday

File: gtdynamics/factors/PoseFactor.h [hunk 2]
```diff
@@ -84,7 +84,8 @@ class PoseFactor : public gtsam::NoiseModelFactor {
     boost::optional<gtsam::Matrix &> wTc_hat_H_q_ref;
     if (H) wTc_hat_H_q_ref = wTc_hat_H_q;
 
-    auto wTc_hat = joint_->poseOf(joint_->child(), wTp, x, t_,
+    auto wTc_hat = joint_->poseOf(joint_->child(), wTp,
+                                  JointAngle<double>(x, joint_->id(), t_),
```
[the remark this answers is not in the document]
Comment (gchenfc) [reply gtdynamics/factors/PoseFactor.h]: Oops I forgot - done

File: gtdynamics/factors/JointMeasurementFactor.h [hunk 3]
```diff
@@ -23,19 +23,15 @@ namespace gtdynamics {
...
 class JointMeasurementFactor
     : public gtsam::NoiseModelFactor2<gtsam::Pose3, gtsam::Pose3> {
  private:
   using This = JointMeasurementFactor;
   using Base = gtsam::NoiseModelFactor2<gtsam::Pose3, gtsam::Pose3>;
 
   JointConstSharedPtr joint_;
-  typename JOINT::JointCoordinate joint_coordinate_;
-  size_t k_;
+  double joint_angle_;
```
[the remark this answers is not in the document]
Comment (gchenfc) [reply gtdynamics/factors/JointMeasurementFactor.h]: done
